File: src/test/java/net/igsoft/sdi/testclasses/PCreator.java
package net.igsoft.sdi.testclasses;

import net.igsoft.sdi.Creator;
import net.igsoft.sdi.InstanceCreator;

public class PCreator extends Creator<P, PCreatorParams> {
    @Override
    public P create(InstanceCreator instanceCreator, PCreatorParams pCreatorParams) {
        R r = instanceCreator.getOrCreate(R.class, new RCreatorParams("name", "surname"));

        return new P(instanceCreator.getOrCreate(Stepper.class), pCreatorParams.getId(), r);
    }
}

File: src/test/scala/net/igsoft/sdi/scala/testclasses/PCreator.scala
package net.igsoft.sdi.scala.testclasses

import net.igsoft.sdi.{Creator, InstanceCreator, ParameterBase}

case class PCreatorParams(manualStartAndStop: Boolean, id: String) extends ParameterBase(manualStartAndStop) {
  override def uniqueId: String = id
}

class PCreator extends Creator[P, PCreatorParams] {
  override def create(instanceCreator: InstanceCreator, pCreatorParams: PCreatorParams): P = {
    val r = instanceCreator.getOrCreate(classOf[R], RCreatorParams("name", "surname"))

    new P(instanceCreator.getOrCreate(classOf[Stepper]), pCreatorParams.id, r)
  }
}
